### PR TITLE
ICFAR-259 Xconf - Telemetry 2.0 - Add Approval Process To 2.0 Profiles

### PR DESCRIPTION
**Pull request by jpethuraj**
Added the approval process features to Telemetry 2.0 profiles.

### CODE REVIEW

File: xconf-angular-admin/src/main/java/com/comcast/xconf/admin/service/telemetrytwochange/TelemetryTwoProfileChangeService.java [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+    @Override
+    public boolean equalPendingEntities(TelemetryTwoProfile oldEntity, TelemetryTwoProfile newEntity) {
+        return true;
+    }
+
+    @Override
+    public List<String> getEntityNames(List<TelemetryTwoChange<TelemetryTwoProfile>> changes) {
+        List<String> names = new ArrayList<>();
+        for (TelemetryTwoChange<TelemetryTwoProfile> change : changes) {
+        	TelemetryTwoProfile profile = TelemetryTwoChangeUtils.getEntity(change);
```
Comment (igor-kostrov) [inline xconf-angular-admin/src/main/java/com/comcast/xconf/admin/service/telemetrytwochange/TelemetryTwoProfileChangeService.java:57]: minor: wrong indentation here.

File: xconf-angular-admin/src/main/java/com/comcast/xconf/admin/controller/telemetry/TelemetryTwoProfileController.java [hunk 2]
```diff
@@ -63,4 +64,24 @@ public String getAllEntitiesExportName() {
     public ResponseEntity getTelemetryTwoProfilesByIdList(@RequestBody List<String> telemetryTwoProfileIdList) {
         return new ResponseEntity<>(telemetryTwoProfileService.getTelemetryTwoProfilesByIdList(telemetryTwoProfileIdList), HttpStatus.OK);
     }
+    
+    @Override
+    @RequestMapping(method = RequestMethod.POST)
+    public ResponseEntity create(@RequestBody TelemetryTwoProfile entity) {
+    	telemetryTwoProfileService.writeCreateChange(entity);
```
Comment (igor-kostrov) [inline xconf-angular-admin/src/main/java/com/comcast/xconf/admin/controller/telemetry/TelemetryTwoProfileController.java:71]: minor: wrong indentation here.

File: xconf-angular-admin/src/main/java/com/comcast/xconf/admin/service/telemetry/TelemetryTwoProfileService.java [hunk 3]
```diff
@@ -92,4 +109,44 @@ protected PermissionService getPermissionService() {
...
+    public TelemetryTwoProfile writeDeleteChange(String id) {
+        beforeRemoving(id);
+        TelemetryTwoProfile profile = getOne(id);
+        pendingChangesService.create(buildToDelete(profile, EntityType.TELEMETRY_TWO_PROFILE, getPermissionService().getWriteApplication(), authService.getUserNameOrUnknown()));
+        return profile;
+    }
+    
+    @Override
+    public TelemetryTwoProfile delete(String id) {
+    	TelemetryTwoProfile delete = super.delete(id);
```
Comment (igor-kostrov) [inline xconf-angular-admin/src/main/java/com/comcast/xconf/admin/service/telemetry/TelemetryTwoProfileService.java:148]: minor: wrong indentation here.

File: xconf-angular-admin/src/main/java/com/comcast/xconf/admin/service/telemetrytwochange/AbstractTelemetryTwoChangeService.java [hunk 4]
```diff
@@ -0,0 +1,197 @@
...
+                if (!changeIdsToBeExcluded.contains(changeByEntityId.getId())) {
+                    changeCrudService.delete(changeByEntityId.getId());
+                    logger.info("Automatically canceled change by {}: {}", authService.getUserName(), changeByEntityId);
+                }
+            }
+        }
+    }
+
+    private void saveToApprovedAndCleanUpChange(TelemetryTwoChange<T> change) {
+    	ApprovedTelemetryTwoChange<T> approvedChange = approvedChangeCrudService.saveToApproved(change);
```
Comment (igor-kostrov) [inline xconf-angular-admin/src/main/java/com/comcast/xconf/admin/service/telemetrytwochange/AbstractTelemetryTwoChangeService.java:103]: minor: wrong indentation here.

File: xconf-angular-admin/src/main/java/com/comcast/xconf/admin/service/telemetrytwochange/AbstractTelemetryTwoChangeService.java [hunk 5]
```diff
@@ -0,0 +1,197 @@
...
+        }
+
+        return updateDeleteEntity(change);
+    }
+
+    private ApprovedTelemetryTwoChange<T> updateDeleteEntity(TelemetryTwoChange<T> change) {
+        T currentEntity = change.getOldEntity();
+        T entityToChange = getEntityService().getOne(change.getEntityId());
+        if (entityToChange != null && equalPendingEntities(currentEntity, entityToChange)) {
+        	ApprovedTelemetryTwoChange<T> approvedChange;
```
Comment (igor-kostrov) [inline xconf-angular-admin/src/main/java/com/comcast/xconf/admin/service/telemetrytwochange/AbstractTelemetryTwoChangeService.java:143]: minor: wrong indentation here.

File: xconf-angular-admin/src/main/java/com/comcast/xconf/admin/service/telemetrytwochange/AbstractTelemetryTwoChangeService.java [hunk 6]
```diff
@@ -0,0 +1,197 @@
...
+            approvedChange = approvedChangeCrudService.saveToApproved(change);
+            changeCrudService.delete(change.getId());
+            return approvedChange;
+        } else {
+            throw new EntityConflictException("Change could not be approved, " + change.getOldEntity().getClass().getSimpleName() + " have been already changed: " + JsonUtil.toJson(entityToChange));
+        }
+    }
+
+    public void revert(String approvedId) {
+    	ApprovedTelemetryTwoChange<T> approvedChange = approvedChangeCrudService.getOne(approvedId);
```
Comment (igor-kostrov) [inline xconf-angular-admin/src/main/java/com/comcast/xconf/admin/service/telemetrytwochange/AbstractTelemetryTwoChangeService.java:160]: minor: wrong indentation here.

File: xconf-angular-admin/src/main/java/com/comcast/xconf/admin/service/telemetrytwochange/AbstractTelemetryTwoChangeService.java [hunk 7]
```diff
@@ -0,0 +1,197 @@
...
+        logger.info("Change has been reverted by {}: {}", authService.getUserName(), approvedChange);
+    }
+
+    public void cancel(String changeId) {
+        TelemetryTwoChange canceledChange = changeCrudService.delete(changeId);
+        logger.info("Change has been canceled by {}: {}", authService.getUserName(), canceledChange);
+    }
+
+    private TelemetryTwoChange<T> revertDelete(String changeId) {
+    	ApprovedTelemetryTwoChange<T> approvedChange = approvedChangeCrudService.getOne(changeId);
```
Comment (igor-kostrov) [inline xconf-angular-admin/src/main/java/com/comcast/xconf/admin/service/telemetrytwochange/AbstractTelemetryTwoChangeService.java:175]: minor: wrong indentation here.

File: xconf-angular-admin/src/main/webapp/app/xconf/telemetrytwochanges/telemetrytwochange.controller.js [hunk 8]
```diff
@@ -0,0 +1,260 @@
...
+        function showSuccessApproveRevertMessage() {
+            if (vm.currentChangeType === vm.CHANGE_TYPE.PENDING) {
+                alertsService.showSuccessMessage({message: 'Selected changes have been successfully approved and unselected changes canceled'});
+            } else {
+                alertsService.showSuccessMessage({message: 'Selected changes were reverted successfully'});
+            }
+        }
+
+        function cleanUpSelectedChanges(selectedChanges, changeIdsWithError) {
+            console.log(changeIdsWithError);
```
Comment (igor-kostrov) [inline xconf-angular-admin/src/main/webapp/app/xconf/telemetrytwochanges/telemetrytwochange.controller.js:229]: Do we need this console.log statement?

File: xconf-angular-admin/src/main/webapp/app/xconf/telemetrytwochanges/telemetrytwochange.controller.js [hunk 9]
```diff
@@ -0,0 +1,260 @@
...
+                if (!changeIdsWithError.includes(change.id)) {
+                    utilsService.removeItemFromListById(selectedChanges, change.id);
+                }
+            });
+        }
+
+        function isAddedToMultipleOperation(change) {
+            for(var i = 0; i < vm.changesForMultipleOperation.length; i++) {
+                if (angular.equals(vm.changesForMultipleOperation[i], change)) {
+                    return true
```
Comment (igor-kostrov) [inline xconf-angular-admin/src/main/webapp/app/xconf/telemetrytwochanges/telemetrytwochange.controller.js:240]: semicolon is missing at the end of statement.

**Comment by igor-kostrov** [comment]
I don't see any changes in tests for TelemetryTwoProfileController. As data flow has changed - and now we have to approve any create/update change - test methods for POST/PUT requests should reflect those changes.

File: xconf-angular-admin/src/main/webapp/app/xconf/telemetrytwochanges/telemetrytwochange.controller.js (hunk 8 above)
Comment (jpethuraj) [reply xconf-angular-admin/src/main/webapp/app/xconf/telemetrytwochanges/telemetrytwochange.controller.js]: removed the console.log

**Comment by jpethuraj** [comment]
> I don't see any changes in tests for TelemetryTwoProfileController. As data flow has changed - and now we have to approve any create/update change - test methods for POST/PUT requests should reflect those changes.

Updated the TelemetryTwoProfileControllerTest for POST/PUT methods

File: xconf-angular-admin/src/main/java/com/comcast/xconf/shared/utils/TelemetryTwoChangeUtils.java [hunk 10]
```diff
@@ -0,0 +1,48 @@
...
+import com.comcast.xconf.change.TelemetryTwoChange;
+
+import org.apache.commons.collections.comparators.NullComparator;
+
+import java.util.Comparator;
+
+public class TelemetryTwoChangeUtils {
+
+    public static <T extends IPersistable> T getEntity(TelemetryTwoChange<T> change) {
+        if (ChangeOperation.CREATE.equals(change.getOperation()) || ChangeOperation.UPDATE.equals(change.getOperation())) {
```
Comment (igor-kostrov) [inline xconf-angular-admin/src/main/java/com/comcast/xconf/shared/utils/TelemetryTwoChangeUtils.java:33]: It's missed IMPORT operation here.

**Review by igor-kostrov — APPROVED** [review]

**Review by comcast-icfar — APPROVED** [review]

**Pull request merged**